### PR TITLE
Add BDD tests for breadcrumbs

### PR DESCRIPTION
**Pull request by slomek**
- go fmt all affected files,
- add missing test scenarios
- refactor access result check function
- add support for partial access in check function

### CODE REVIEW

File: features/get-all-items-breadcrumbs.feature [hunk 1]
```diff
@@ -1,56 +1,247 @@
...
-And the response body should be, in JSON:
-  """
-  [
-    { "item_id": 21, "language_id": 1, "title": "Graph: Methods" },
-    { "item_id": 22, "language_id": 1, "title": "DFS" },
-    { "item_id": 23, "language_id": 1, "title": "Reduce Graph" },
-    { "item_id": 21, "language_id": 2, "title": "Graphe: Methodes" }
-  ]
-  """
+  Scenario: All items from a valid breadcrumb
```
[the remark this answers is not in the document]
Comment (smadbe) [reply features/get-all-items-breadcrumbs.feature]: btw, all the scenario name should describe (in summary) the expected outcode, e.g. "should fail with 'unauthorized'", ... not just the starting behavior. Help quick browsing of the scenario to understand the expected behavior.

**Comment by smadbe** [comment]
For the testcov, actually as you "open" the tests for a new package, it is normal indeed.

File: app/database/itemStore.go [hunk 2]
```diff
@@ -1,106 +1,125 @@
...
+		return false, nil
+	}
+	return true, nil
+}
+
+// checkAccess checks if the user has access to all items:
+// - user has to have full access to all items
+// OR
+// - user has to have full access to all but last, and grayed access to that last item.
+func checkAccess(itemIDs []int64, accRes []accessResult) error {
```
[the remark this answers is not in the document]
Comment (slomek) [reply app/database/itemStore.go]: 😊

File: app/database/itemStore_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package database
+
+import (
+	"fmt"
+	"testing"
+)
+
+func Test_checkAccess(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (slomek) [reply app/database/itemStore_test.go]: It doesn't, I use that to show that I'm testing an unexported function, but I'm fine with changing it to the more standard one.

File: app/database/itemStore.go [hunk 4]
```diff
@@ -1,106 +1,125 @@
...
-    if !found {
-      logging.Logger.Infof("User access validation failed: not visible item_id %d", id)
-      return false, nil // no row matching this item_id
-    }
-  }
-
-  return true, nil
+	var accRes []accessResult
+	db := s.GroupItems().MatchingUserAncestors(user).
+		Select("idItem, MAX(bCachedFullAccess) AS fullAccess, MAX(bCachedPartialAccess) AS partialAccess, MAX(bCachedGrayedAccess) AS grayedAccess").
```
[the remark this answers is not in the document]
Comment (slomek) [reply app/database/itemStore.go]: Not sure what you mean, there is no "alignment with spaces" in Go.

File: app/database/itemStore.go [hunk 5]
```diff
@@ -1,106 +1,125 @@
...
-      }
-    }
-    if !found {
-      logging.Logger.Infof("User access validation failed: not visible item_id %d", id)
-      return false, nil // no row matching this item_id
-    }
-  }
-
-  return true, nil
+	var accRes []accessResult
```
[the remark this answers is not in the document]
Comment (slomek) [reply app/database/itemStore.go]: Sure, I'll change it to `itemAccessDetails`.

File: features/get-all-items-breadcrumbs.feature [hunk 6]
```diff
@@ -1,56 +1,247 @@
...
+      [
+      { "item_id": 21, "language_id": 1, "title": "Graph: Methods" },
+      { "item_id": 22, "language_id": 1, "title": "DFS" },
+      { "item_id": 23, "language_id": 1, "title": "Reduce Graph" },
+      { "item_id": 21, "language_id": 2, "title": "Graphe: Methodes" }
+      ]
+      """
+
+  Scenario: Partial access to all items except for last which is greyed
+    Given the database has the following table 'users':
```
[the remark this answers is not in the document]
Comment (slomek) [reply features/get-all-items-breadcrumbs.feature]: I'm not sure how to do that, do you have any experience how to do that?

File: features/get-all-items-breadcrumbs.feature (hunk 1 above)
[the remark this answers is not in the document]
Comment (slomek) [reply features/get-all-items-breadcrumbs.feature]: I've changed them a bit, let me know if they are OK. If not, please suggest the names :)

File: features/get-all-items-breadcrumbs.feature [hunk 7]
```diff
@@ -1,56 +1,247 @@
...
+      """
+      [
+      { "item_id": 21, "language_id": 1, "title": "Graph: Methods" },
+      { "item_id": 22, "language_id": 1, "title": "DFS" },
+      { "item_id": 23, "language_id": 1, "title": "Reduce Graph" },
+      { "item_id": 21, "language_id": 2, "title": "Graphe: Methodes" }
+      ]
+      """
+
+  Scenario: Partial access to all items except for last which is greyed
```
[the remark this answers is not in the document]
Comment (slomek) [reply features/get-all-items-breadcrumbs.feature]: I saw just one test case, I can add another if necessary.

**Comment by slomek** [comment]
@smadbe I've updated the PR, let me know what you think.

File: features/get-all-items-breadcrumbs.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply features/get-all-items-breadcrumbs.feature]: Actually, for now the whole context is resetted for each scenario (`steps.go:28`) while we should keep what is defined at the feature level for each scenario. Let's do this later.

**Pull request merged**